### ISSUE TITLE
optimise G_BotThink() by reducing the loop and sharing the result for all bots

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#2118]
Currently, each bot iterate through the whole list of entities in game at each call to G_BotThink().

This implies many calculations which are repeated, notably distance calculations (which can even be calculated more than once for a single G_BotThink() call).
As I guess all bots do their G_BotThink() sequentially, it is quite unlikely that any entity moved in between, so the results could be shared to avoid many calculations and iterations.

Also, they loop through all entities.
This implies map-entities and missiles, notably. While this might be interesting in distant future to have bots dodge missiles, fire on ground, moving entities, avoid trigger_hurt, benefit from trigger_heal, etc, we are far from being able to do that right now, so those iterations are basically lost cycles.

**Comment by slipher** [comment]
Yes, most loops over all entities are actually only interested in a few of them and should be using `ForEntities`.

**Comment by bmorel** [comment]
From what I read, it's possible this would avoid iterating over useless entities yes. If I understand it correctly, this mess is maintained on each Entity creation/deletion. This is still a std::set, usually implemented with a RBtree of pointers to elements though, and there's not that many entities to iterate over, so  I'm not sure this would actually improve on the performance problem considering the fact iterating entities is linear.
But if better performance can be reached with that, it's worth a try to use this overcomplicated and underdocumented thing.
